modules/backendscheduler/provider/retention.go

```diff
@@ -0,0 +1,86 @@
+package provider
+
+import (
+	"context"
+	"flag"
+	"time"
+
+	kitlogger "github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/google/uuid"
+	"github.com/grafana/tempo/modules/backendscheduler/work"
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+type RetentionConfig struct {
+	Interval time.Duration `yaml:"interval"`
+}
+
+func (cfg *RetentionConfig) RegisterFlagsAndApplyDefaults(prefix string, f *flag.FlagSet) {
+	f.DurationVar(&cfg.Interval, prefix+"backend-scheduler.retention-interval", time.Hour, "Interval at which to perform tenant retention")
+}
+
+type RetentionProvider struct {
+	cfg    RetentionConfig
+	sched  Scheduler
+	logger kitlogger.Logger
+}
+
+func NewRetentionProvider(cfg RetentionConfig, logger kitlogger.Logger, scheduler Scheduler) *RetentionProvider {
+	return &RetentionProvider{
+		cfg:    cfg,
+		sched:  scheduler,
+		logger: logger,
+	}
+}
+
+func (p *RetentionProvider) Start(ctx context.Context) <-chan *work.Job {
+	jobs := make(chan *work.Job, 1)
+
+	go func() {
+		defer close(jobs)
+		ticker := time.NewTicker(p.cfg.Interval)
+		defer ticker.Stop()
+
+		level.Info(p.logger).Log("msg", "retention provider started")
+
+		for {
+			select {
+			case <-ctx.Done():
+				level.Info(p.logger).Log("msg", "retention provider stopping")
+				return
+			case <-ticker.C:
+				if job := p.nextRetentionJob(); job != nil {
+					select {
+					case jobs <- job:
+					default:
+						// Channel full, try again next tick
+					}
+				}
+			}
+		}
+	}()
+
+	return jobs
+}
+
+func (p *RetentionProvider) nextRetentionJob() *work.Job {
+	// Check if we already have a retention job running
+	for _, j := range p.sched.ListJobs() {
```

Comment:
i'm unsure on putting this here. imo this kind of "control" or "business" logic should be handled entirely in the backendscheduler. this would give us one place to look to see how jobs are prioritized/scheduled.

i think this is fine for now, but as we expand the number of providers, job types etc, i believe we will see a need to consolidate this kind of logic in one spot.

Reply:
I see this also.  Let's see if we can clean this up with the next provider we add.